File: src/main/java/org/cascadebot/cascadebot/data/objects/TodoList.java
package org.cascadebot.cascadebot.data.objects;

import lombok.Getter;
import lombok.Setter;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.TextChannel;
import org.cascadebot.cascadebot.CascadeBot;
import org.cascadebot.cascadebot.UnicodeConstants;
import org.cascadebot.cascadebot.commandmeta.CommandContext;
import org.cascadebot.cascadebot.data.managers.GuildDataManager;
import org.cascadebot.cascadebot.data.objects.guild.GuildData;
import org.cascadebot.cascadebot.messaging.Messaging;
import org.cascadebot.cascadebot.messaging.MessagingObjects;
import org.cascadebot.cascadebot.utils.buttons.PersistentButton;
import org.cascadebot.cascadebot.utils.buttons.PersistentButtonGroup;

import java.util.ArrayList;
import java.util.List;

public class TodoList {


    @Getter
    private List<TodoListItem> items = new ArrayList<>();

    @Getter
    @Setter
    private long messageId = -1;

    @Getter
    @Setter
    private long channelId = -1;

    @Getter
    @Setter
    private int currentItem;

    @Getter
    private long ownerId;

    //List of users id who are able to access this list
    private List<Long> users = new ArrayList<>();

    private TodoList() {
        //Constructor for mongodb
    }

    public TodoList(long ownerId) {
        this.ownerId = ownerId;
    }

    public int addTodoItem(String text) {
        TodoListItem item = new TodoListItem(text);
        items.add(item);
        return items.indexOf(item);
    }

    public TodoListItem removeTodoItem(int id) {
        return items.remove(id);
    }

    public void addEditUser(Member member) {
        users.add(member.getIdLong());
    }

    public void removeEditUser(Member member) {
        users.remove(member.getIdLong());
    }

    public boolean canUserEdit(Long id) {
        return ownerId == id || users.contains(id);
    }

    public static class TodoListItem {

        @Getter
        @Setter
        private boolean done;

        @Getter
        @Setter
        private String text;

        private TodoListItem() {
            //Constructor for mongodb
        }

        TodoListItem(String text) {
            this.text = text;
            done = false;
        }

    }

    public void edit(CommandContext context) {
        if (messageId == -1 || channelId == -1) return;
        TextChannel originalChannel = context.getGuild().getTextChannelById(channelId);
        if (originalChannel != null && originalChannel.getIdLong() == channelId) {
            Message message = originalChannel.retrieveMessageById(messageId).complete();
            if (message != null) {
                message.editMessage(getTodoListMessage()).queue();
                doCheckToggle(message);
            }
        }
    }

    public void send(CommandContext context, TextChannel channel) {
        if (channel == null) {
            throw new IllegalArgumentException("The channel should exist :(");
        }
        PersistentButtonGroup buttonGroup = generateButtons(context.getMember().getIdLong(), channel.getIdLong(), context.getGuild().getIdLong());
        this.setCurrentItem(0);
        Messaging.sendButtonedMessage(channel, getTodoListMessage(), buttonGroup).thenAccept(message -> {
            messageId = message.getIdLong();
            this.channelId = message.getChannel().getIdLong();
        });
    }

    private PersistentButtonGroup generateButtons(long memberId, long channelId, long guildId) {
        PersistentButtonGroup buttonGroup = new PersistentButtonGroup(memberId, channelId, guildId);
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_LEFT);
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_UP);
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_DOWN);
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_RIGHT);
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_CHECK);
        return buttonGroup;
    }

    public void addUncheckButton(Message message) {
        TextChannel channel = CascadeBot.INS.getClient().getTextChannelById(channelId);
        if (channel != null) {
            GuildData data = GuildDataManager.getGuildData(channel.getGuild().getIdLong());
            PersistentButtonGroup buttonGroup = data.getPersistentButtons().get(channelId).get(messageId);
            buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_UNCHECK);
            buttonGroup.removePersistentButton(PersistentButton.TODO_BUTTON_CHECK);
        }
    }

    public void addCheckButton(Message message) {
        TextChannel channel = CascadeBot.INS.getClient().getTextChannelById(channelId);
        if (channel != null) {
            GuildData data = GuildDataManager.getGuildData(channel.getGuild().getIdLong());
            PersistentButtonGroup buttonGroup = data.getPersistentButtons().get(channelId).get(messageId);
            buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_CHECK);
            buttonGroup.removePersistentButton(PersistentButton.TODO_BUTTON_UNCHECK);
        }
    }

    public void doCheckToggle(Message message) {
        TodoList.TodoListItem item = items.get(this.getCurrentItem());

        if (item.isDone()) {
            addUncheckButton(message);
        } else {
            addCheckButton(message);
        }
        //TODO check all items check. I'm going to wait for persistent buttons to do this
    }

    public MessageEmbed getTodoListMessage() {
        int pos = this.getCurrentItem();
        int currentPage = pos / 10 + 1;
        int start = currentPage * 10 - 10;
        int end = start + 9;

        StringBuilder pageBuilder = new StringBuilder();

        for (int i = start; i <= end; i++) {
            if (i >= this.getItems().size()) {
                break;
            }
            TodoList.TodoListItem item = this.getItems().get(i);
            if (i == pos) {
                pageBuilder.append(UnicodeConstants.SMALL_ORANGE_DIAMOND).append(" ");
            } else {
                pageBuilder.append(UnicodeConstants.WHITE_SMALL_SQUARE).append(" ");
            }
            pageBuilder.append(i + 1).append(": ");
            if (item.isDone()) {
                pageBuilder.append("~~");
            }
            pageBuilder.append(item.getText()).append('\n');
            if (item.isDone()) {
                pageBuilder.append("~~");
            }
        }
        EmbedBuilder embedBuilder = MessagingObjects.getClearThreadLocalEmbedBuilder();
        embedBuilder.setTitle("Todo List");
        embedBuilder.appendDescription(pageBuilder.toString());
        return embedBuilder.build();
    }

}

File: src/main/kotlin/org/cascadebot/cascadebot/data/objects/TodoList.kt
package org.cascadebot.cascadebot.data.objects

import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.TextChannel
import org.bson.BsonDocument
import org.cascadebot.cascadebot.CascadeBot
import org.cascadebot.cascadebot.UnicodeConstants
import org.cascadebot.cascadebot.commandmeta.CommandContext
import org.cascadebot.cascadebot.data.database.BsonObject
import org.cascadebot.cascadebot.data.managers.GuildDataManager
import org.cascadebot.cascadebot.messaging.Messaging.sendButtonedMessage
import org.cascadebot.cascadebot.messaging.MessagingObjects
import org.cascadebot.cascadebot.utils.GuildDataUtils.assertWriteMode
import org.cascadebot.cascadebot.utils.buttons.PersistentButton
import org.cascadebot.cascadebot.utils.buttons.PersistentButtonGroup

class TodoList(val ownerId: Long) : BsonObject {

    val items: MutableList<TodoListItem> = ArrayList()
    var messageId: Long = -1
    var channelId: Long = -1
    var currentItem = 0

    //List of users id who are able to access this list
    private val users: MutableList<Long> = ArrayList()

    private constructor() : this(0) {
        //Constructor for mongodb
    }

    fun addTodoItem(text: String?): Int {
        assertWriteMode()
        val item = TodoListItem(text)
        items.add(item)
        return items.indexOf(item)
    }

    fun removeTodoItem(id: Int): TodoListItem {
        assertWriteMode()
        return items.removeAt(id)
    }

    fun addEditUser(member: Member) {
        assertWriteMode()
        users.add(member.idLong)
    }

    fun removeEditUser(member: Member) {
        assertWriteMode()
        users.remove(member.idLong)
    }

    fun canUserEdit(id: Long): Boolean {
        return ownerId == id || users.contains(id)
    }

    class TodoListItem {

        

        var done = false
        var text: String? = null

        private constructor() {
            //Constructor for mongodb
        }

        internal constructor(text: String?) {
            this.text = text
            done = false
        }
    }

    fun edit(context: CommandContext) {
        assertWriteMode()
        if (messageId == -1L || channelId == -1L) return
        val originalChannel = context.guild.getTextChannelById(channelId)
        if (originalChannel != null && originalChannel.idLong == channelId) {
            val message = originalChannel.retrieveMessageById(messageId).complete()
            if (message != null) {
                message.editMessage(todoListMessage).queue()
                doCheckToggle(message)
            }
        }
    }

    fun send(context: CommandContext, channel: TextChannel?) {
        requireNotNull(channel) { "The channel should exist :(" }
        val buttonGroup = generateButtons(context.member.idLong, channel.idLong, context.guild.idLong)
        currentItem = 0
        sendButtonedMessage(channel, todoListMessage, buttonGroup).thenAccept {
            messageId = it.idLong
            channelId = it.channel.idLong
        }
    }

    private fun generateButtons(memberId: Long, channelId: Long, guildId: Long): PersistentButtonGroup {
        val buttonGroup = PersistentButtonGroup(memberId, channelId, guildId)
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_LEFT)
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_UP)
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_DOWN)
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_NAVIGATE_RIGHT)
        buttonGroup.addPersistentButton(PersistentButton.TODO_BUTTON_CHECK)
        return buttonGroup
    }

    fun addUncheckButton(message: Message?) {
        assertWriteMode()
        val channel = CascadeBot.INS.client.getTextChannelById(channelId)
        if (channel != null) {
            val data = GuildDataManager.getGuildData(channel.guild.idLong)
            val buttonGroup = data.persistentButtons[channelId]!![messageId]
            buttonGroup!!.addPersistentButton(PersistentButton.TODO_BUTTON_UNCHECK)
            buttonGroup.removePersistentButton(PersistentButton.TODO_BUTTON_CHECK)
        }
    }

    fun addCheckButton(message: Message?) {
        assertWriteMode()
        val channel = CascadeBot.INS.client.getTextChannelById(channelId)
        if (channel != null) {
            val data = GuildDataManager.getGuildData(channel.guild.idLong)
            val buttonGroup = data.persistentButtons[channelId]!![messageId]
            buttonGroup!!.addPersistentButton(PersistentButton.TODO_BUTTON_CHECK)
            buttonGroup.removePersistentButton(PersistentButton.TODO_BUTTON_UNCHECK)
        }
    }

    fun doCheckToggle(message: Message?) {
        assertWriteMode()
        val item = items[this.currentItem]
        if (item.done) {
            addUncheckButton(message)
        } else {
            addCheckButton(message)
        }
        //TODO check all items check. I'm going to wait for persistent buttons to do this
    }

    val todoListMessage: MessageEmbed
        get() {
            val pos: Int = this.currentItem
            val currentPage = pos / 10 + 1
            val start = currentPage * 10 - 10
            val end = start + 9
            val pageBuilder = StringBuilder()
            for (i in start..end) {
                if (i >= this.items.size) {
                    break
                }
                val item: TodoListItem = this.items.get(i)
                if (i == pos) {
                    pageBuilder.append(UnicodeConstants.SMALL_ORANGE_DIAMOND).append(" ")
                } else {
                    pageBuilder.append(UnicodeConstants.WHITE_SMALL_SQUARE).append(" ")
                }
                pageBuilder.append(i + 1).append(": ")
                if (item.done) {
                    pageBuilder.append("~~")
                }
                pageBuilder.append(item.text).append('\n')
                if (item.done) {
                    pageBuilder.append("~~")
                }
            }
            val embedBuilder = MessagingObjects.getClearThreadLocalEmbedBuilder()
            embedBuilder.setTitle("Todo List")
            embedBuilder.appendDescription(pageBuilder.toString())
            return embedBuilder.build()
        }

    override fun fromBson(bsonDocument: BsonDocument) {
        if (bsonDocument.contains("channelId")) {
            channelId = bsonDocument["channelId"]!!.asNumber().longValue()
        }
        if (bsonDocument.contains("messageId")) {
            messageId = bsonDocument["messageId"]!!.asNumber().longValue()
        }
        if (bsonDocument.contains("currentItem")) {
            currentItem = bsonDocument["currentItem"]!!.asNumber().intValue()
        }
        if (bsonDocument.contains("items")) {
            items.clear();
            for (item in bsonDocument["items"]!!.asArray()) {
                val todoItem = TodoListItem(item.asDocument()["text"]!!.asString().value)
                todoItem.done = item.asDocument()["done"]!!.asBoolean().value
                items.add(todoItem)
            }
        }
        CascadeBot.INS.client.getTextChannelById(channelId)!!.editMessageById(messageId, todoListMessage).queue()
    }
}
